pkg/api/server/v1alpha2/log/s3.go

```diff
@@ -0,0 +1,149 @@
+package log
+
+import (
+	"bufio"
+	"errors"
+	"path/filepath"
+	"strings"
+
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/tektoncd/results/pkg/apis/v1alpha2"
+	"github.com/tektoncd/results/pkg/conf"
+)
+
+type S3LogStreamer struct {
+	LogStreamer
+	io.WriterTo
+	io.ReaderFrom
+
+	bufSize int
+	conf    *conf.ConfigFile
+	path string
+}
+
+func NewS3LogStreamer(trl *v1alpha2.TaskRunLog, bufSize int, conf *conf.ConfigFile, logDataDir string) LogStreamer {
+	if trl.Status.File == nil {
+		trl.Status.File = &v1alpha2.FileLogTypeStatus{
+			Path: defaultFilePath(trl),
+		}
+	}
+	return &S3LogStreamer{
+		conf:    conf,
+		bufSize: bufSize,
+		path: filepath.Join(logDataDir, trl.Status.File.Path),
+	}
+}
+
+func (*S3LogStreamer) Type() string {
+	return string(v1alpha2.S3LogType)
+}
+
+func (ls *S3LogStreamer) WriteTo(w io.Writer) (n int64, err error) {
+	client := ls.initConfig()
+
+	outPut, err := client.GetObject(context.TODO(), &s3.GetObjectInput{
+		Bucket: &ls.conf.S3_BUCKET_NAME,
+		Key:    &ls.path,
+	})
+	if err != nil {
+		return 0, fmt.Errorf(err.Error())
+	}
+
+	defer outPut.Body.Close()
+
+	reader := bufio.NewReaderSize(outPut.Body, ls.bufSize)
+	n, err = reader.WriteTo(w)
+	if err != nil {
+		return 0, fmt.Errorf(err.Error())
+	}
+	return
+}
+
+func (ls *S3LogStreamer) ReadFrom(r io.Reader) (n int64, err error) {
+	client := ls.initConfig()
+
+	prevContent, err := ls.ReadS3(client)
+	if err != nil {
+		var oe *types.NoSuchKey
+		if errors.As(err, &oe) {
+			prevContent = []byte{}
+		} else {
+			return 0, fmt.Errorf(err.Error())
+		}
+	}
+
+	newBts, err := io.ReadAll(r)
+	if err != nil {
+		return 0, fmt.Errorf(err.Error())
+	}
+
+	_, err = client.PutObject(context.TODO(), &s3.PutObjectInput{
```

Comment:
I recommend adding a `context` object to the S3 streamer. We should use a passed in context, which helps with timeouts, cancellation, and so forth.